File: pyScripts/create_Y_E_from_icdlab.py
"""
Create Y and E tensors from existing icdlab RDS file
Adapted from sample_for_RAP.py to work with pre-processed icdlab data
"""

import pandas as pd
import numpy as np
import torch
import rpy2.robjects as ro
from rpy2.robjects import pandas2ri
from rpy2.robjects.packages import importr

# Activate pandas conversion
pandas2ri.activate()

# Load RDS file
print("="*80)
print("Loading icdlab RDS file...")
print("="*80)

base = importr('base')
icdlab = base.readRDS("~/Dropbox-Personal/icd10phe_lab.rds")

# Convert to pandas DataFrame
df = pandas2ri.rpy2py(icdlab)
print(f"Loaded {len(df):,} rows")
print(f"Columns: {list(df.columns)}")
print(f"Unique patients: {df.iloc[:, 0].nunique():,}")

# Assume columns are: patient_id, age_diag, phecode (diag_icd10)
# Adjust column names if different
col_names = list(df.columns)
patient_col = col_names[0]  # First column should be patient ID
age_col = [c for c in col_names if 'age' in c.lower()][0] if any('age' in c.lower() for c in col_names) else col_names[1]
phecode_col = [c for c in col_names if any(x in c.lower() for x in ['phecode', 'icd', 'diag'])][0] if any(any(x in c.lower() for x in ['phecode', 'icd', 'diag']) for c in col_names) else col_names[2]

print(f"\nUsing columns:")
print(f"  Patient: {patient_col}")
print(f"  Age: {age_col}")
print(f"  Phecode: {phecode_col}")

# Rename for consistency
df = df.rename(columns={
    patient_col: 'eid',
    age_col: 'age_diag',
    phecode_col: 'diag_icd10'
})

# Filter age range (29-80)
print(f"\nFiltering to age 29-80...")
df = df[df['age_diag'].between(29, 80, inclusive='both')].copy()
df['age_idx'] = (df['age_diag'].round().astype(int) - 29)
df = df[df['age_idx'] >= 0]  # Ensure non-negative

print(f"After filtering: {len(df):,} rows")

# Integer encode
print(f"\nCreating integer encodings...")
eids = df['eid'].astype(str).unique()
phecodes = df['diag_icd10'].astype(str).unique()

eid_index = {eid: idx for idx, eid in enumerate(eids)}
phe_index = {code: idx for idx, code in enumerate(phecodes)}

n_patients = len(eids)
n_diseases = len(phecodes)
n_timepoints = df['age_idx'].max() + 1

print(f"Patients: {n_patients:,}")
print(f"Diseases (phecodes): {n_diseases:,}")
print(f"Timepoints: {n_timepoints}")

# Create Y tensor
print(f"\nCreating Y tensor...")
Y = torch.zeros((n_patients, n_diseases, n_timepoints), dtype=torch.int8)

# Fill Y tensor
print("Filling Y tensor (this may take a while)...")
for idx, row in df.iterrows():
    eid_str = str(row['eid'])
    phecode_str = str(row['diag_icd10'])
    age_idx = int(row['age_idx'])
    
    if eid_str in eid_index and phecode_str in phe_index:
        Y[eid_index[eid_str], phe_index[phecode_str], age_idx] = 1
    
    if (idx + 1) % 100000 == 0:
        print(f"  Processed {idx + 1:,} / {len(df):,} rows")

print("✓ Y tensor created")

# Create E tensor (event times)
print(f"\nCreating E tensor...")
def create_event_matrix(Y_tensor):
    n_patients, n_diseases, n_times = Y_tensor.shape
    max_time = n_times - 1
    E = torch.full((n_patients, n_diseases), max_time, dtype=torch.int16)
    
    events = torch.nonzero(Y_tensor == 1, as_tuple=False)
    for event in events:
        patient, disease, time = event[0].item(), event[1].item(), event[2].item()
        if time < E[patient, disease].item():
            E[patient, disease] = time
    
    return E

E = create_event_matrix(Y)
print("✓ E tensor created")

# Save outputs
print(f"\nSaving outputs...")
output_dir = "~/Dropbox-Personal/data_for_running/"
import os
output_dir = os.path.expanduser(output_dir)
os.makedirs(output_dir, exist_ok=True)

torch.save(Y, os.path.join(output_dir, "Y_tensor.pt"))
torch.save(E, os.path.join(output_dir, "E_matrix.pt"))

# Also save as numpy format
np.savez_compressed(
    os.path.join(output_dir, "Y_tensor.npz"),
    Y=Y.numpy()
)
np.savez_compressed(
    os.path.join(output_dir, "E_matrix.npz"),
    E=E.numpy()
)

# Save metadata
import pickle
metadata = {
    'n_patients': n_patients,
    'n_diseases': n_diseases,
    'n_timepoints': n_timepoints,
    'eids': eids,
    'phecodes': phecodes,
    'eid_index': eid_index,
    'phecode_index': phe_index
}

with open(os.path.join(output_dir, "Y_E_metadata.pkl"), 'wb') as f:
    pickle.dump(metadata, f)

print("="*80)
print("COMPLETE")
print("="*80)
print(f"Y tensor shape: {Y.shape}")
print(f"E tensor shape: {E.shape}")
print(f"Output directory: {output_dir}")
print(f"\nFiles saved:")
print(f"  - Y_tensor.pt")
print(f"  - E_matrix.pt")
print(f"  - Y_tensor.npz")
print(f"  - E_matrix.npz")
print(f"  - Y_E_metadata.pkl")
print("="*80)

File: pyScripts/create_Y_E_from_icdlab.R
#--- Setup -------------------------------------------------
library(data.table)
library(dplyr)
library(torch)

#--- Load existing ICD/Phecode data ------------------------
icdlab <- readRDS("~/Dropbox-Personal/icd10phe_lab.rds")

# Convert to data.table for efficiency
if (!is.data.table(icdlab)) {
  icdlab <- as.data.table(icdlab)
}

# Check column names (adjust if needed)
print("Column names:")
print(names(icdlab))
print(paste("Total rows:", nrow(icdlab)))
print(paste("Unique patients:", length(unique(icdlab[[1]]))))

# Assume columns are: patient_id, age_diag, phecode (diag_icd10)
# Adjust column names if different
col_names <- names(icdlab)
patient_col <- col_names[1]  # First column should be patient ID
age_col <- col_names[grep("age", col_names, ignore.case = TRUE)[1]]  # Age column
phecode_col <- col_names[grep("phecode|icd|diag", col_names, ignore.case = TRUE)[1]]  # Phecode column

print(paste("Using columns:"))
print(paste("  Patient:", patient_col))
print(paste("  Age:", age_col))
print(paste("  Phecode:", phecode_col))

# Rename for consistency
icdlab <- icdlab %>%
  rename(
    eid = !!sym(patient_col),
    age_diag = !!sym(age_col),
    diag_icd10 = !!sym(phecode_col)
  )

#--- Filter age range (29-80) ----------------------------
icdlab_filtered <- icdlab %>%
  filter(age_diag >= 29, age_diag <= 80) %>%
  mutate(age_idx = round(age_diag) - 29) %>%
  filter(age_idx >= 0)  # Ensure non-negative

print(paste("After filtering (age 29-80):", nrow(icdlab_filtered), "rows"))

#--- Create integer encodings ------------------------------
eids <- unique(icdlab_filtered$eid)
phecodes <- unique(icdlab_filtered$diag_icd10)

eid_index <- setNames(seq_along(eids) - 1, eids)  # 0-indexed for Python
phecode_index <- setNames(seq_along(phecodes) - 1, phecodes)  # 0-indexed for Python

n_patients <- length(eids)
n_diseases <- length(phecodes)
n_timepoints <- max(icdlab_filtered$age_idx) + 1

print(paste("Patients:", n_patients))
print(paste("Diseases (phecodes):", n_diseases))
print(paste("Timepoints:", n_timepoints))

#--- Create Y tensor --------------------------------------
Y <- torch_zeros(c(n_patients, n_diseases, n_timepoints), dtype = torch_int8())

print("Filling Y tensor...")
pb <- txtProgressBar(min = 0, max = nrow(icdlab_filtered), style = 3)

for (i in seq_len(nrow(icdlab_filtered))) {
  row <- icdlab_filtered[i]
  eid_idx <- eid_index[[as.character(row$eid)]]
  phecode_idx <- phecode_index[[as.character(row$diag_icd10)]]
  age_idx <- row$age_idx + 1  # R is 1-indexed
  
  if (!is.na(eid_idx) && !is.na(phecode_idx) && !is.na(age_idx)) {
    Y[eid_idx + 1, phecode_idx + 1, age_idx] <- 1L  # R is 1-indexed
  }
  
  if (i %% 10000 == 0) {
    setTxtProgressBar(pb, i)
  }
}
close(pb)

#--- Create E tensor (event times) -----------------------
print("Creating E tensor...")
create_event_matrix <- function(Y_tensor) {
  n_patients <- Y_tensor$size(1)
  n_diseases <- Y_tensor$size(2)
  n_times <- Y_tensor$size(3)
  max_time <- n_times - 1
  
  E <- torch_full(c(n_patients, n_diseases), max_time, dtype = torch_int16())
  
  # Find all events
  events <- torch_nonzero(Y_tensor == 1, as_tuple = FALSE)
  
  if (events$numel() > 0) {
    for (i in seq_len(events$size(1))) {
      patient <- events[i, 1]$item() + 1  # Convert to R 1-indexed
      disease <- events[i, 2]$item() + 1
      time <- events[i, 3]$item() + 1
      
      if (time < E[patient, disease]$item()) {
        E[patient, disease] <- (time - 1)  # Store as 0-indexed
      }
    }
  }
  
  return(E)
}

E <- create_event_matrix(Y)

#--- Save outputs -----------------------------------------
output_dir <- "~/Dropbox-Personal/data_for_running/"
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

torch_save(Y, file.path(output_dir, "Y_tensor.pt"))
torch_save(E, file.path(output_dir, "E_matrix.pt"))

# Also save as numpy format (for Python compatibility)
Y_np <- Y$numpy()
E_np <- E$numpy()

# Save using reticulate or write directly
library(reticulate)
np <- import("numpy")
np$savez_compressed(
  file.path(output_dir, "Y_tensor.npz"),
  Y = Y_np
)
np$savez_compressed(
  file.path(output_dir, "E_matrix.npz"),
  E = E_np
)

# Save metadata
metadata <- list(
  n_patients = n_patients,
  n_diseases = n_diseases,
  n_timepoints = n_timepoints,
  eids = eids,
  phecodes = phecodes,
  eid_index = eid_index,
  phecode_index = phecode_index
)

saveRDS(metadata, file.path(output_dir, "Y_E_metadata.rds"))

print("="*80)
print("COMPLETE")
print("="*80)
print(paste("Y tensor shape:", paste(Y$shape, collapse = " x ")))
print(paste("E tensor shape:", paste(E$shape, collapse = " x ")))
print(paste("Output directory:", output_dir))
print(paste("Files saved:"))
print(paste("  - Y_tensor.pt"))
print(paste("  - E_matrix.pt"))
print(paste("  - Y_tensor.npz"))
print(paste("  - E_matrix.npz"))
print(paste("  - Y_E_metadata.rds"))

